src/Lean/Message.lean

```diff
@@ -78,12 +78,38 @@ inductive MessageData where
     Example: an inspector that tries to find "definitional equality failures" may look for the tag "DefEqFailure". -/
   | tagged            : Name → MessageData → MessageData
   | trace (data : TraceData) (msg : MessageData) (children : Array MessageData)
-  deriving Inhabited
+  /-- Lazy message data production. The `Dynamic` is expected to be `MessageData`, which
+  is a workaround for the positivity restriction. -/
+  | ofLazy (f : Option PPContext → IO Dynamic) (hasSyntheticSorry : MetavarContext → Bool)
+  deriving Inhabited, TypeName
 
 namespace MessageData
 
+/--
+Lazy message data production, with access to the context as given by
+a surrounding `MessageData.withContext` (which is expected to exist).
+-/
+def lazy (f : PPContext → IO MessageData) (hasSyntheticSorry : MetavarContext → Bool) : MessageData :=
+  .ofLazy (hasSyntheticSorry := hasSyntheticSorry) fun ctx? => do
+    let msg ← match ctx? with
+      | .none => pure (.ofFormat "(invalid MessageData.lazy, missing context)")
+      | .some ctx => f ctx
+    return Dynamic.mk msg
+
+/-- Lazily formatted text with Info annotations -/
+def ofPPFormat (f : PPFormat) : MessageData :=
```

Comment:
`PPFormat` was introduced to unify various lazy formatters, and is hardly used in the codebase. As this PR introduces a more general lazy constructor, I suggest that `ofPPFormat` and `PPFormat` be removed to clean up the API.

Reply:
I have now carried out this experiment [here](https://github.com/Vtec234/lean4/tree/msgdata-refactor). The fallout in core appears minimal.

Reply:
Thanks, looks good! Do you want to push that to my branch?

Reply:
Done!

Reply:
Thanks! Now mathlib fails, and searching for the error message on the phone is annoying. But probably nothing worrying and we can merge?